tests/test_schema.py

```diff
@@ -1411,3 +1412,43 @@ class Model(BaseModel):
             }
         },
     }
+
+
+def test_new_type_schema():
+    a_type = NewType('a_type', int)
+    b_type = NewType('b_type', a_type)
+    c_type = NewType('c_type', str)
+
+    class Model(BaseModel):
+        a: a_type
+        b: b_type
+        c: c_type
+
+    assert Model.schema() == {
+        'properties': {
+            'a': {'title': 'A', 'type': 'integer'},
+            'b': {'title': 'B', 'type': 'integer'},
+            'c': {'title': 'C', 'type': 'string'},
+        },
+        'required': ['a', 'b', 'c'],
+        'title': 'Model',
+        'type': 'object',
+    }
+
+
+def test_literal_schema():
+    class Model(BaseModel):
+        a: Literal[1]
+        b: Literal['a']
+        c: Literal['a', 1]
+
+    assert Model.schema() == {
+        'properties': {
+            'a': {'title': 'A', 'type': 'integer'},
+            'b': {'title': 'B', 'type': 'string'},
+            'c': {'anyOf': [{'type': 'integer'}, {'type': 'string'}], 'title': 'C'},
```

Comment:
I suppose it might be preferable for this to look more like the schema for an Enum, but I'm not sure what is right. Let me know if we prefer something else.

Reply:
This should be `enum`:

```suggestion
            'c': {'enum': ['a', 1], 'title': 'C'},
```

Reply:
@samuelcolvin would you agree on this point? If so, I'll modify to ensure it gets `'enum'` here

Reply:
1. 'anyOf' is wrong because in JSON Schema parlance 'anyOf' means it should validate against _one or more_ subschemas in the array. You want to use 'oneOf'.
2. 'type' is wrong because it should match one of the values in the array exactly, meaning you'd have to use 'const'.
3. `'oneOf': [{'const': 'a'}, {'const': 1}]` is a roundabout way to say `'enum': ['a', 1]`.

Reply:
I'm not a JSONSchema expert either, @tiangolo what do you think?

Reply:
@layday Thanks for the info; I am convinced it should either be enum or oneOf. I'm inclined to go with enum here, because I think that is the most useful external interpretation, but there is a subtle issue that might argue in favor of using oneOf with const:

The [mypy docs](https://mypy.readthedocs.io/en/latest/literal_types.html) currently say:
> Future versions of mypy may relax some of these restrictions. For example, we plan on adding support for using enum values inside Literal[...] in an upcoming release.

So I take that to mean that right now, you can't do `Literal[MyEnum.case]`. So, it would actually break mypy if you were to actually use an enum in the Literal in the python code. Obviously you can work around this by just not using an enum in your code, but it sort of feels internally inconsistent if direct interpretation of the schema is not compatible with the code used to generate it. In particular, this could pose an annoyance to efforts to automatically generate a pydantic model from a schema.

Given this, it seemed to me that it could make sense to use `'oneOf': [{'const': 'a'}, {'const': 1}]` to reflect the use of `Literal`, and `enum` to signify the use of an `Enum`. On the other hand, there are obvious benefits to just using enum externally.

I'll probably just convert it to enum unless I hear objections, but let me know if you think this point might be relevant @layday @samuelcolvin

Reply:
I don't think it's entirely reasonable to expect that you should be able to map schemas injectively to type hints in Python and vice versa.  `Enum`s in particular cannot be directly translated to JSON Schema.  Your `Literal[MyEnum.case]` would be exported as `{"const": <whatever the case may be>}` and how would you know it came from an `Enum`?  If you were to export an `Enum` to an `enum` in the schema, how would you go about rebuilding the `Enum`?  It's far more likely that you'd opt for a `Union` or a multi-value `Literal`.

Reply:
I have used openapi-generator with FastAPI and it is capable of generating full-fledged enums based on a pydantic model with an enum property based on the values. At any rate, I'll interpret your comment as a rejection of the alternative.

Reply:
And what does it use for `Enum` member names?

Reply:
I use the same name as value in my enums (specifically so generation works nicely), and the enum member names are just the values. (I just tried with integers, and the generated cases are just named `_0`, `_1` etc.; not very useful.)

In a generation context, I think the expectation has to be that the enum name and value are the same for this reason (or at least closely related, e.g. snake-vs-camelcase).

Reply:
I was imagining a scenario where the values were strings; if you want to use int literals, it could be harder to interpret.

The openapi-generated enums use the same name as the value; I use the same enum name as value everywhere.

Reply:
This is why I suggested unions or literals might be preferable.  JSON Schema `enum`s can contain any JSON type, including null, booleans, arrays and objects.

Reply:
I would suggest `oneOf` instead of `enum`.

That would allow having extra data in the schema. For example, if you use a `NewType` based on `EmailStr`, the `type` would be `string` with `format: "email"`. But if it's an `enum`, it won't be able to have the extra `format: "email"`.

Also, `const` can go along with `type`. `type` could be used by clients (e.g. frontends) to display the value in a certain way, even if it can only be a specific value.

E.g.: 

`'oneOf': [{'type': 'string', 'const': 'a'}, {'type': 'integer', 'const': 1}]`

Reply:
Literals can only contain primitive literals - not `NewType` - so the situation you describe does not arise.  But if you think including the `type` can be useful let's go with `oneOf`.

Reply:
I have updated this to use `'const'`, but I refactored to use `Union` to handle multiple values. This is consistent with the semantics of `Literal`, (i.e., `Literal[1, 2, 3]` is the same as `Union[Literal[1], Literal[2], Literal[3]]`), and means that we can leverage the Union-handling logic to generate a multi-value schema.

Currently, this results in `anyOf` rather than `oneOf`. In the case of `Literal`, it is impossible to match on multiple, so `anyOf` is essentially equivalent to `oneOf`. I figured this was preferable to implementing custom handling of the multi-value case, but could rework to explicitly use `oneOf` if there is a good reason.

Alternatively, https://github.com/samuelcolvin/pydantic/issues/656 might result in Union using `oneOf` anyway, in which case the point would be moot anyway.

Reply:
Great! Agree.

And actually, after thinking about it, I guess `anyOf` makes more sense, as I guess there might be cases where even with `Literal`, a single value could be valid with more than one schema. Not sure, but in any case, it is not enforcing that a value *not* valid against the other schemas, so, it's not really enforcing a `oneOf`.

Reply:
In short, agreed.